api/metric/api_test.go

```diff
@@ -0,0 +1,698 @@
+		handle := c.GetHandle(labels)
+		handle.Add(ctx, 42)
+		meter.RecordBatch(ctx, labels, c.Measurement(42))
+		checkBatches(t, ctx, labels, meter, c.Descriptor())
+	}
+}
+
+func TestGauge(t *testing.T) {
+	{
+		g := NewFloat64Gauge("ajwaj")
+		meter := newMockMeter()
+		ctx := context.Background()
+		labels := meter.DefineLabels(ctx)
+		g.Set(ctx, 42, labels)
+		handle := g.GetHandle(labels)
+		handle.Set(ctx, 42)
+		meter.RecordBatch(ctx, labels, g.Measurement(42))
+		checkBatches(t, ctx, labels, meter, g.Descriptor())
+	}
+	{
+		g := NewInt64Gauge("ajwaj")
+		meter := newMockMeter()
+		ctx := context.Background()
+		labels := meter.DefineLabels(ctx)
+		g.Set(ctx, 42, labels)
+		handle := g.GetHandle(labels)
+		handle.Set(ctx, 42)
+		meter.RecordBatch(ctx, labels, g.Measurement(42))
+		checkBatches(t, ctx, labels, meter, g.Descriptor())
+	}
+}
+
+func TestMeasure(t *testing.T) {
+	{
+		m := NewFloat64Measure("ajwaj")
+		meter := newMockMeter()
+		ctx := context.Background()
+		labels := meter.DefineLabels(ctx)
+		m.Record(ctx, 42, labels)
+		handle := m.GetHandle(labels)
+		handle.Record(ctx, 42)
+		meter.RecordBatch(ctx, labels, m.Measurement(42))
+		checkBatches(t, ctx, labels, meter, m.Descriptor())
+	}
+	{
+		m := NewInt64Measure("ajwaj")
+		meter := newMockMeter()
+		ctx := context.Background()
+		labels := meter.DefineLabels(ctx)
+		m.Record(ctx, 42, labels)
+		handle := m.GetHandle(labels)
+		handle.Record(ctx, 42)
+		meter.RecordBatch(ctx, labels, m.Measurement(42))
+		checkBatches(t, ctx, labels, meter, m.Descriptor())
+	}
+}
+
+func TestObserver(t *testing.T) {
+	{
+		o := NewFloat64Observer("ajwaj")
+		meter := newMockMeter()
+		ctx := context.Background()
+		labels := meter.DefineLabels(ctx)
+		RegisterFloat64Observer(meter, o, func(meter Meter, o Float64Observer, cb Float64ObservationCallback) {
+			cb(labels, 42)
+			cb(labels, 42)
+			cb(labels, 42)
+		})
+		meter.PerformObservations()
+		checkBatches(t, ctx, labels, meter, o.Descriptor())
+	}
+	{
+		o := NewInt64Observer("ajwaj")
+		ctx := context.Background()
+		meter := newMockMeter()
+		labels := meter.DefineLabels(ctx)
+		RegisterInt64Observer(meter, o, func(meter Meter, o Int64Observer, cb Int64ObservationCallback) {
+			cb(labels, 42)
+			cb(labels, 42)
+			cb(labels, 42)
+		})
+		meter.PerformObservations()
+		checkBatches(t, ctx, labels, meter, o.Descriptor())
+	}
+}
+
+func checkBatches(t *testing.T, ctx context.Context, labels LabelSet, meter *mockMeter, descriptor *Descriptor) {
+	if len(meter.measurementBatches) != 3 {
+		t.Errorf("Expected 3 recorded measurement batches, got %d", len(meter.measurementBatches))
+	}
+	minLen := 3
+	if minLen > len(meter.measurementBatches) {
+		minLen = len(meter.measurementBatches)
+	}
+	for i := 0; i < minLen; i++ {
+		got := meter.measurementBatches[i]
+		if got.ctx != ctx {
+			t.Errorf("Wrong recorded context in batch %d", i)
+		}
+		if got.labelSet != labels {
+			t.Errorf("Wrong recorded label set in batch %d", i)
```

Comment:
Just printing batch index may not be sufficient when a test fails.
Same comment for all error messages.

Given that there will be changes in the API as per new proposal, you may ignore this comment.

Reply:
+1 on both counts

Reply:
Done.